Product filter

kart-c opened the pull request:
## Add product filters 

### Changes made in this PR
- Implement price slider - on the initial render slider value is set to maximum. 
- Implement category filter - User can select multiple categories and products will filter accordingly. If no categories are selected all products will be displayed to the user.
- Implement Rating filter - User can choose to view products depending on the product rating.
- Implement Sort By price - Users can sort products depending on their price (ascending or descending)
- Implement Availability - Users can choose not to display out-of-stock products.
- If there are no products to display a simple `No Products found` message is displayed for now.
- Users can clear all filters by using the `Clear` button on the filter section.

[CSB for refernce](https://08i3yg.csb.app/)

kart-c committed:
refactor: refactor style of card and product grid

remove 100% width of card and give fixed width for each grid column

kart-c committed:
feat: create global filter context

kart-c committed:
feat: add filter reducer

kart-c committed:
feat: add filter functions in utils and export them

kart-c committed:
refactor: export context and reducer using default index js

kart-c committed:
feat: add filter functionality in filters.jsx

kart-c committed:
feat: add filter functions in products and use them

display final products array. if length is less than 0 show message to user

kart-c committed:
feat: show user current value of slider

Kevin-Solomon commented on src/Context/Filter Context/filter-context.js:
```diff
@@ -0,0 +1,26 @@
+import { createContext, useContext, useReducer } from 'react';
+import { filterReducer } from '../../Reducer';
+
+const FilterContext = createContext({
+	price: 5000,
+	category: [],
+	rating: 1,
+	sortBy: null,
+	outOfStock: false,
+});
```
Just a small suggestion, you can also move the initial state in the util file.

kart-c replied:
Since the code is only of 5 lines, I think it is okay for. Thank you for the suggestion tho @Kevin-Solomon 👍

Kevin-Solomon commented on src/Pages/Products/Components/Filter/Filters.jsx:
```diff
@@ -1,64 +1,177 @@
 import styles from './Filters.module.css';
 
 const Filters = () => {
+	const { state, dispatch } = useFilter();
```
Since we are going to have multiple contexts and reducers you can probably rename ```dispatch``` to ```filterDispatch```

kart-c replied:
Thanks, I will change that right away Kevin 🚀

RahullRawat reviewed (approved):
- Functionality is working fine
- Good that you have created separate functions for different categories.

AtulPant2704 reviewed (approved):
Great naming convention, was really easy to understand. Just one change which Kevin has already mentioned regarding naming of state and dispatch.

AnkurChunekar reviewed (approved):
- I liked the fact that you have included an arrray for categories.
- code is very neat and clean
- reducer and context in separate files made it easy to review, Great work @kart-c

kart-c committed:
refactor: rename state and dispatch as, filterState and FilterDispatch respectively

kart-c committed:
fix: fix clear button

when user clicked on clear filters, remove-out-of-stock checkbox was not cleared. Fixed the issue
